{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import numpy as np\n",
    "\n",
    "from scquint.differential_splicing import run_differential_splicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 44518 × 29965\n",
       "    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden', 'cell_type', 'plate_id'\n",
       "    var: 'chromosome', 'start', 'end', 'cluster', 'left_exon_count', 'left_genes', 'right_exon_count', 'right_genes', 'genes', 'n_genes', 'genes_cluster', 'n_genes_cluster', 'annotated', 'gene_id', 'original_cluster', 'n_start', 'strand', 'cluster_type'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Tabula Muris intron count matrix can be downloaded at \n",
    "# https://figshare.com/articles/dataset/scQuint_data_objects_-_Tabula_Muris/14471904?file=27696714\n",
    "adata = anndata.read_h5ad(\"/home/gbenegas/Desktop/tabula_muris/adata_spl.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "fibroblast of cardiac tissue           2118\n",
       "endothelial cell of coronary artery    1179\n",
       "monocyte                                490\n",
       "endocardial cell                        164\n",
       "atrial myocyte                          159\n",
       "valve cell                              158\n",
       "smooth muscle cell                       65\n",
       "ventricular myocyte                      46\n",
       "T cell                                   37\n",
       "B cell                                   15\n",
       "Name: cell_ontology_class, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs.tissue==\"Heart\"].cell_ontology_class.value_counts()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164, 159)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify the indices of the two groups to be compared\n",
    "cell_idx_a = np.where((adata.obs.tissue==\"Heart\") & (adata.obs.cell_ontology_class==\"endocardial cell\"))[0]\n",
    "cell_idx_b = np.where((adata.obs.tissue==\"Heart\") & (adata.obs.cell_ontology_class==\"atrial myocyte\"))[0]\n",
    "len(cell_idx_a), len(cell_idx_b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.var` of view, copying.\n",
      "Trying to set attribute `.var` of view, copying.\n",
      "Trying to set attribute `.var` of view, copying.\n",
      "Trying to set attribute `.var` of view, copying.\n",
      "Trying to set attribute `.var` of view, copying.\n",
      "Trying to set attribute `.var` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample sizes:  164 159\n",
      "filter_min_cells_per_feature\n",
      "filter_singletons\n",
      "filter_min_cells_per_cluster\n",
      "filter_singletons\n",
      "filter_min_cells_per_cluster\n",
      "filter_singletons\n",
      "Number of intron clusters:  84\n",
      "Number of introns:  169\n",
      "Testing intron cluster  0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gbenegas/anaconda3/lib/python3.6/site-packages/torch/autograd/__init__.py:132: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)\n",
      "  allow_unreachable=True)  # allow_unreachable flag\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "clusters, introns = run_differential_splicing(\n",
    "    adata, cell_idx_a, cell_idx_b, min_cells_per_cluster=50, min_total_cells_per_intron=50, n_jobs=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_value</th>\n",
       "      <th>ll_null</th>\n",
       "      <th>ll</th>\n",
       "      <th>n_classes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>max_abs_delta_psi</th>\n",
       "      <th>max_abs_lfc_psi</th>\n",
       "      <th>ranking</th>\n",
       "      <th>p_value_adj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1269</th>\n",
       "      <td>4.198450e-47</td>\n",
       "      <td>-358.7117348210924</td>\n",
       "      <td>-254.82372878310215</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000061904</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>8.692238</td>\n",
       "      <td>0</td>\n",
       "      <td>3.526698e-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12347</th>\n",
       "      <td>5.336710e-41</td>\n",
       "      <td>-304.4631111680536</td>\n",
       "      <td>-214.55895498416908</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000032366</td>\n",
       "      <td>0.797576</td>\n",
       "      <td>3.716576</td>\n",
       "      <td>1</td>\n",
       "      <td>2.241418e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6211</th>\n",
       "      <td>4.635299e-39</td>\n",
       "      <td>-452.0730904495759</td>\n",
       "      <td>-366.60815780882535</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000025781</td>\n",
       "      <td>0.717983</td>\n",
       "      <td>2.737905</td>\n",
       "      <td>2</td>\n",
       "      <td>1.297884e-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11316</th>\n",
       "      <td>1.163422e-36</td>\n",
       "      <td>-250.61955418336956</td>\n",
       "      <td>-170.64721919349435</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000008892</td>\n",
       "      <td>0.792927</td>\n",
       "      <td>5.902056</td>\n",
       "      <td>3</td>\n",
       "      <td>2.443186e-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15076</th>\n",
       "      <td>1.653410e-30</td>\n",
       "      <td>-118.71946751177697</td>\n",
       "      <td>-52.81864338786248</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000018845</td>\n",
       "      <td>0.893422</td>\n",
       "      <td>5.195932</td>\n",
       "      <td>4</td>\n",
       "      <td>2.777729e-29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              p_value              ll_null                   ll  n_classes  \\\n",
       "cluster                                                                      \n",
       "1269     4.198450e-47   -358.7117348210924  -254.82372878310215          2   \n",
       "12347    5.336710e-41   -304.4631111680536  -214.55895498416908          2   \n",
       "6211     4.635299e-39   -452.0730904495759  -366.60815780882535          2   \n",
       "11316    1.163422e-36  -250.61955418336956  -170.64721919349435          2   \n",
       "15076    1.653410e-30  -118.71946751177697   -52.81864338786248          2   \n",
       "\n",
       "                    gene_id  max_abs_delta_psi  max_abs_lfc_psi  ranking  \\\n",
       "cluster                                                                    \n",
       "1269     ENSMUSG00000061904           0.798296         8.692238        0   \n",
       "12347    ENSMUSG00000032366           0.797576         3.716576        1   \n",
       "6211     ENSMUSG00000025781           0.717983         2.737905        2   \n",
       "11316    ENSMUSG00000008892           0.792927         5.902056        3   \n",
       "15076    ENSMUSG00000018845           0.893422         5.195932        4   \n",
       "\n",
       "          p_value_adj  \n",
       "cluster                \n",
       "1269     3.526698e-45  \n",
       "12347    2.241418e-39  \n",
       "6211     1.297884e-37  \n",
       "11316    2.443186e-35  \n",
       "15076    2.777729e-29  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clusters has information about all the intron clusters that were tested\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>cluster</th>\n",
       "      <th>left_exon_count</th>\n",
       "      <th>left_genes</th>\n",
       "      <th>right_exon_count</th>\n",
       "      <th>right_genes</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>original_cluster</th>\n",
       "      <th>n_start</th>\n",
       "      <th>strand</th>\n",
       "      <th>cluster_type</th>\n",
       "      <th>psi_a</th>\n",
       "      <th>psi_b</th>\n",
       "      <th>delta_psi</th>\n",
       "      <th>lfc_psi</th>\n",
       "      <th>abs_delta_psi</th>\n",
       "      <th>abs_lfc_psi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>91119707</td>\n",
       "      <td>91121922</td>\n",
       "      <td>1269</td>\n",
       "      <td>9</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>5</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1269</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.998065</td>\n",
       "      <td>0.199769</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>2.320800</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>2.320800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>91119707</td>\n",
       "      <td>91122207</td>\n",
       "      <td>1269</td>\n",
       "      <td>9</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>2</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1269</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.001935</td>\n",
       "      <td>0.800231</td>\n",
       "      <td>-0.798296</td>\n",
       "      <td>-8.692238</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>8.692238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>128491034</td>\n",
       "      <td>128491720</td>\n",
       "      <td>1417</td>\n",
       "      <td>11</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>1</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1417</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.765274</td>\n",
       "      <td>0.207626</td>\n",
       "      <td>0.557649</td>\n",
       "      <td>1.881993</td>\n",
       "      <td>0.557649</td>\n",
       "      <td>1.881993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>128491034</td>\n",
       "      <td>128492059</td>\n",
       "      <td>1417</td>\n",
       "      <td>11</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>14</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>['ENSMUSG00000090841']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1417</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.234726</td>\n",
       "      <td>0.792374</td>\n",
       "      <td>-0.557649</td>\n",
       "      <td>-1.755205</td>\n",
       "      <td>0.557649</td>\n",
       "      <td>1.755205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>4702221</td>\n",
       "      <td>4703904</td>\n",
       "      <td>1472</td>\n",
       "      <td>4</td>\n",
       "      <td>['ENSMUSG00000059534']</td>\n",
       "      <td>2</td>\n",
       "      <td>['ENSMUSG00000059534']</td>\n",
       "      <td>['ENSMUSG00000059534']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1472</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.033252</td>\n",
       "      <td>0.104812</td>\n",
       "      <td>-0.071559</td>\n",
       "      <td>-1.656266</td>\n",
       "      <td>0.071559</td>\n",
       "      <td>1.656266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome      start        end  cluster  left_exon_count  \\\n",
       "0      chr10   91119707   91121922     1269                9   \n",
       "1      chr10   91119707   91122207     1269                9   \n",
       "2      chr10  128491034  128491720     1417               11   \n",
       "3      chr10  128491034  128492059     1417               11   \n",
       "4      chr11    4702221    4703904     1472                4   \n",
       "\n",
       "               left_genes  right_exon_count             right_genes  \\\n",
       "0  ['ENSMUSG00000061904']                 5  ['ENSMUSG00000061904']   \n",
       "1  ['ENSMUSG00000061904']                 2  ['ENSMUSG00000061904']   \n",
       "2  ['ENSMUSG00000090841']                 1  ['ENSMUSG00000090841']   \n",
       "3  ['ENSMUSG00000090841']                14  ['ENSMUSG00000090841']   \n",
       "4  ['ENSMUSG00000059534']                 2  ['ENSMUSG00000059534']   \n",
       "\n",
       "                    genes  n_genes  ... original_cluster  n_start  strand  \\\n",
       "0  ['ENSMUSG00000061904']        1  ...             1269        1       -   \n",
       "1  ['ENSMUSG00000061904']        1  ...             1269        1       -   \n",
       "2  ['ENSMUSG00000090841']        1  ...             1417        1       -   \n",
       "3  ['ENSMUSG00000090841']        1  ...             1417        1       -   \n",
       "4  ['ENSMUSG00000059534']        1  ...             1472        1       -   \n",
       "\n",
       "      cluster_type     psi_a     psi_b delta_psi   lfc_psi  abs_delta_psi  \\\n",
       "0  shared_acceptor  0.998065  0.199769  0.798296  2.320800       0.798296   \n",
       "1  shared_acceptor  0.001935  0.800231 -0.798296 -8.692238       0.798296   \n",
       "2  shared_acceptor  0.765274  0.207626  0.557649  1.881993       0.557649   \n",
       "3  shared_acceptor  0.234726  0.792374 -0.557649 -1.755205       0.557649   \n",
       "4  shared_acceptor  0.033252  0.104812 -0.071559 -1.656266       0.071559   \n",
       "\n",
       "   abs_lfc_psi  \n",
       "0     2.320800  \n",
       "1     8.692238  \n",
       "2     1.881993  \n",
       "3     1.755205  \n",
       "4     1.656266  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# introns has information about all the introns corresponding to the tested clusters\n",
    "introns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of significant differential splicing events\n",
    "(clusters.p_value_adj < 0.05).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>cluster</th>\n",
       "      <th>left_exon_count</th>\n",
       "      <th>left_genes</th>\n",
       "      <th>right_exon_count</th>\n",
       "      <th>right_genes</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>original_cluster</th>\n",
       "      <th>n_start</th>\n",
       "      <th>strand</th>\n",
       "      <th>cluster_type</th>\n",
       "      <th>psi_a</th>\n",
       "      <th>psi_b</th>\n",
       "      <th>delta_psi</th>\n",
       "      <th>lfc_psi</th>\n",
       "      <th>abs_delta_psi</th>\n",
       "      <th>abs_lfc_psi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>91119707</td>\n",
       "      <td>91121922</td>\n",
       "      <td>1269</td>\n",
       "      <td>9</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>5</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1269</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.998065</td>\n",
       "      <td>0.199769</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>2.320800</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>2.320800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>91119707</td>\n",
       "      <td>91122207</td>\n",
       "      <td>1269</td>\n",
       "      <td>9</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>2</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>['ENSMUSG00000061904']</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1269</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>shared_acceptor</td>\n",
       "      <td>0.001935</td>\n",
       "      <td>0.800231</td>\n",
       "      <td>-0.798296</td>\n",
       "      <td>-8.692238</td>\n",
       "      <td>0.798296</td>\n",
       "      <td>8.692238</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome     start       end  cluster  left_exon_count  \\\n",
       "0      chr10  91119707  91121922     1269                9   \n",
       "1      chr10  91119707  91122207     1269                9   \n",
       "\n",
       "               left_genes  right_exon_count             right_genes  \\\n",
       "0  ['ENSMUSG00000061904']                 5  ['ENSMUSG00000061904']   \n",
       "1  ['ENSMUSG00000061904']                 2  ['ENSMUSG00000061904']   \n",
       "\n",
       "                    genes  n_genes  ... original_cluster  n_start  strand  \\\n",
       "0  ['ENSMUSG00000061904']        1  ...             1269        1       -   \n",
       "1  ['ENSMUSG00000061904']        1  ...             1269        1       -   \n",
       "\n",
       "      cluster_type     psi_a     psi_b delta_psi   lfc_psi  abs_delta_psi  \\\n",
       "0  shared_acceptor  0.998065  0.199769  0.798296  2.320800       0.798296   \n",
       "1  shared_acceptor  0.001935  0.800231 -0.798296 -8.692238       0.798296   \n",
       "\n",
       "   abs_lfc_psi  \n",
       "0     2.320800  \n",
       "1     8.692238  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The introns belonging to the intron cluster with the lowest p-value\n",
    "introns[introns.cluster==1269]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The results can be visualized with \n",
    "# the cellxgene browser https://scquint.ds.czbiohub.org/tabula-muris/\n",
    "# and UCSC Genome Browser https://genome.ucsc.edu/s/gbenegas/tabulamuris"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
