{
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "# Install scQuint\n",
        "!pip install -U git+https://github.com/songlab-cal/scquint.git"
      ],
      "metadata": {
        "id": "ZqEj_KcGSpyV",
        "outputId": "ab5ad42c-5d18-48e9-d723-b774b0d9b4e4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting git+https://github.com/songlab-cal/scquint.git\n",
            "  Cloning https://github.com/songlab-cal/scquint.git to /tmp/pip-req-build-jzxqivup\n",
            "  Running command git clone -q https://github.com/songlab-cal/scquint.git /tmp/pip-req-build-jzxqivup\n",
            "Collecting anndata\n",
            "  Downloading anndata-0.8.0-py3-none-any.whl (96 kB)\n",
            "\u001b[K     |████████████████████████████████| 96 kB 3.0 MB/s \n",
            "\u001b[?25hCollecting scanpy\n",
            "  Downloading scanpy-1.9.1-py3-none-any.whl (2.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.0 MB 41.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: networkx in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (2.6.3)\n",
            "Requirement already satisfied: torch in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.11.0+cu113)\n",
            "Collecting pyro-ppl\n",
            "  Downloading pyro_ppl-1.8.1-py3-none-any.whl (718 kB)\n",
            "\u001b[K     |████████████████████████████████| 718 kB 44.6 MB/s \n",
            "\u001b[?25hRequirement already satisfied: scikit-learn in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.0.2)\n",
            "Collecting snakemake\n",
            "  Downloading snakemake-7.8.1.tar.gz (302 kB)\n",
            "\u001b[K     |████████████████████████████████| 302 kB 71.7 MB/s \n",
            "\u001b[?25hRequirement already satisfied: statsmodels in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (0.10.2)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.1.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.21.6)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.3.5)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.7/dist-packages (from scquint==0.2) (1.4.1)\n",
            "Requirement already satisfied: packaging>=20 in /usr/local/lib/python3.7/dist-packages (from anndata->scquint==0.2) (21.3)\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.7/dist-packages (from anndata->scquint==0.2) (5.5.0)\n",
            "Requirement already satisfied: importlib_metadata>=0.7 in /usr/local/lib/python3.7/dist-packages (from anndata->scquint==0.2) (4.11.4)\n",
            "Requirement already satisfied: h5py>=3 in /usr/local/lib/python3.7/dist-packages (from anndata->scquint==0.2) (3.1.0)\n",
            "Requirement already satisfied: typing_extensions in /usr/local/lib/python3.7/dist-packages (from anndata->scquint==0.2) (4.2.0)\n",
            "Requirement already satisfied: cached-property in /usr/local/lib/python3.7/dist-packages (from h5py>=3->anndata->scquint==0.2) (1.5.2)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib_metadata>=0.7->anndata->scquint==0.2) (3.8.0)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20->anndata->scquint==0.2) (3.0.9)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->scquint==0.2) (2022.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas->scquint==0.2) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas->scquint==0.2) (1.15.0)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.7/dist-packages (from pyro-ppl->scquint==0.2) (3.3.0)\n",
            "Requirement already satisfied: tqdm>=4.36 in /usr/local/lib/python3.7/dist-packages (from pyro-ppl->scquint==0.2) (4.64.0)\n",
            "Collecting pyro-api>=0.1.1\n",
            "  Downloading pyro_api-0.1.2-py3-none-any.whl (11 kB)\n",
            "Collecting session-info\n",
            "  Downloading session_info-1.0.0.tar.gz (24 kB)\n",
            "Requirement already satisfied: patsy in /usr/local/lib/python3.7/dist-packages (from scanpy->scquint==0.2) (0.5.2)\n",
            "Collecting umap-learn>=0.3.10\n",
            "  Downloading umap-learn-0.5.3.tar.gz (88 kB)\n",
            "\u001b[K     |████████████████████████████████| 88 kB 7.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numba>=0.41.0 in /usr/local/lib/python3.7/dist-packages (from scanpy->scquint==0.2) (0.51.2)\n",
            "Collecting matplotlib>=3.4\n",
            "  Downloading matplotlib-3.5.2-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (11.2 MB)\n",
            "\u001b[K     |████████████████████████████████| 11.2 MB 39.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: seaborn in /usr/local/lib/python3.7/dist-packages (from scanpy->scquint==0.2) (0.11.2)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=3.4->scanpy->scquint==0.2) (1.4.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=3.4->scanpy->scquint==0.2) (0.11.0)\n",
            "Collecting fonttools>=4.22.0\n",
            "  Downloading fonttools-4.33.3-py3-none-any.whl (930 kB)\n",
            "\u001b[K     |████████████████████████████████| 930 kB 56.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=3.4->scanpy->scquint==0.2) (7.1.2)\n",
            "Requirement already satisfied: llvmlite<0.35,>=0.34.0.dev0 in /usr/local/lib/python3.7/dist-packages (from numba>=0.41.0->scanpy->scquint==0.2) (0.34.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from numba>=0.41.0->scanpy->scquint==0.2) (57.4.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->scquint==0.2) (3.1.0)\n",
            "Collecting pynndescent>=0.5\n",
            "  Downloading pynndescent-0.5.7.tar.gz (1.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.1 MB 50.5 MB/s \n",
            "\u001b[?25hCollecting stdlib_list\n",
            "  Downloading stdlib_list-0.8.0-py3-none-any.whl (63 kB)\n",
            "\u001b[K     |████████████████████████████████| 63 kB 2.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: wrapt in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (1.14.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (2.23.0)\n",
            "Collecting ratelimiter\n",
            "  Downloading ratelimiter-1.2.0.post0-py3-none-any.whl (6.6 kB)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (3.13)\n",
            "Collecting configargparse\n",
            "  Downloading ConfigArgParse-1.5.3-py3-none-any.whl (20 kB)\n",
            "Requirement already satisfied: appdirs in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (1.4.4)\n",
            "Collecting datrie\n",
            "  Downloading datrie-0.8.2-cp37-cp37m-manylinux1_x86_64.whl (537 kB)\n",
            "\u001b[K     |████████████████████████████████| 537 kB 59.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: jsonschema in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (4.3.3)\n",
            "Requirement already satisfied: docutils in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (0.17.1)\n",
            "Collecting gitpython\n",
            "  Downloading GitPython-3.1.27-py3-none-any.whl (181 kB)\n",
            "\u001b[K     |████████████████████████████████| 181 kB 73.7 MB/s \n",
            "\u001b[?25hRequirement already satisfied: psutil in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (5.4.8)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (5.4.0)\n",
            "Collecting toposort\n",
            "  Downloading toposort-1.7-py2.py3-none-any.whl (9.0 kB)\n",
            "Collecting connection_pool>=0.0.3\n",
            "  Downloading connection_pool-0.0.3.tar.gz (3.8 kB)\n",
            "Collecting pulp>=2.0\n",
            "  Downloading PuLP-2.6.0-py3-none-any.whl (14.2 MB)\n",
            "\u001b[K     |████████████████████████████████| 14.2 MB 50.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: smart_open>=3.0 in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (6.0.0)\n",
            "Collecting stopit\n",
            "  Downloading stopit-1.1.2.tar.gz (18 kB)\n",
            "Requirement already satisfied: tabulate in /usr/local/lib/python3.7/dist-packages (from snakemake->scquint==0.2) (0.8.9)\n",
            "Collecting yte<2.0,>=1.0\n",
            "  Downloading yte-1.5.1-py3-none-any.whl (7.5 kB)\n",
            "Collecting jinja2<4.0,>=3.0\n",
            "  Downloading Jinja2-3.1.2-py3-none-any.whl (133 kB)\n",
            "\u001b[K     |████████████████████████████████| 133 kB 58.0 MB/s \n",
            "\u001b[?25hCollecting retry\n",
            "  Downloading retry-0.9.2-py2.py3-none-any.whl (8.0 kB)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.7/dist-packages (from jinja2<4.0,>=3.0->snakemake->scquint==0.2) (2.0.1)\n",
            "Collecting pyyaml\n",
            "  Downloading PyYAML-6.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (596 kB)\n",
            "\u001b[K     |████████████████████████████████| 596 kB 36.5 MB/s \n",
            "\u001b[?25hCollecting dpath<3.0,>=2.0\n",
            "  Downloading dpath-2.0.6-py3-none-any.whl (15 kB)\n",
            "Collecting plac<2.0.0,>=1.3.4\n",
            "  Downloading plac-1.3.5-py2.py3-none-any.whl (22 kB)\n",
            "Collecting gitdb<5,>=4.0.1\n",
            "  Downloading gitdb-4.0.9-py3-none-any.whl (63 kB)\n",
            "\u001b[K     |████████████████████████████████| 63 kB 1.7 MB/s \n",
            "\u001b[?25hCollecting smmap<6,>=3.0.1\n",
            "  Downloading smmap-5.0.0-py3-none-any.whl (24 kB)\n",
            "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema->snakemake->scquint==0.2) (0.18.1)\n",
            "Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema->snakemake->scquint==0.2) (21.4.0)\n",
            "Requirement already satisfied: importlib-resources>=1.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema->snakemake->scquint==0.2) (5.7.1)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.7/dist-packages (from nbformat->snakemake->scquint==0.2) (4.10.0)\n",
            "Requirement already satisfied: traitlets>=5.1 in /usr/local/lib/python3.7/dist-packages (from nbformat->snakemake->scquint==0.2) (5.1.1)\n",
            "Requirement already satisfied: fastjsonschema in /usr/local/lib/python3.7/dist-packages (from nbformat->snakemake->scquint==0.2) (2.15.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->snakemake->scquint==0.2) (2022.5.18.1)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->snakemake->scquint==0.2) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->snakemake->scquint==0.2) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->snakemake->scquint==0.2) (1.24.3)\n",
            "Requirement already satisfied: decorator>=3.4.2 in /usr/local/lib/python3.7/dist-packages (from retry->snakemake->scquint==0.2) (4.4.2)\n",
            "Requirement already satisfied: py<2.0.0,>=1.4.26 in /usr/local/lib/python3.7/dist-packages (from retry->snakemake->scquint==0.2) (1.11.0)\n",
            "Building wheels for collected packages: scquint, umap-learn, pynndescent, session-info, snakemake, connection-pool, stopit\n",
            "  Building wheel for scquint (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for scquint: filename=scquint-0.2-py3-none-any.whl size=8920 sha256=d8d382cf93e18357ab2191ed49e46cb3b9263ab7feacee6cf0a13c3ba2a6d6a9\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-01axtvju/wheels/60/7a/bd/1dacc7a9d14b27f14a92016878aebcf4ce97a532d46a31d20b\n",
            "  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for umap-learn: filename=umap_learn-0.5.3-py3-none-any.whl size=82829 sha256=2d9f7a5c4176153e23bbf3c26341608622982edcbb8bc2a9fa85c74f9ad9ceb4\n",
            "  Stored in directory: /root/.cache/pip/wheels/b3/52/a5/1fd9e3e76a7ab34f134c07469cd6f16e27ef3a37aeff1fe821\n",
            "  Building wheel for pynndescent (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pynndescent: filename=pynndescent-0.5.7-py3-none-any.whl size=54286 sha256=532e9b7d4a315ed145eb9fb8795f1aafe8138f299adf7a30616125e7c6066253\n",
            "  Stored in directory: /root/.cache/pip/wheels/7f/2a/f8/7bd5dcec71bd5c669f6f574db3113513696b98f3f9b51f496c\n",
            "  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8048 sha256=ae2655fedcb8cef4a22b27599ddbec52775c36ddbd56a2807620d07655a15bd5\n",
            "  Stored in directory: /root/.cache/pip/wheels/bd/ad/14/6a42359351a18337a8683854cfbba99dd782271f2d1767f87f\n",
            "  Building wheel for snakemake (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for snakemake: filename=snakemake-7.8.1-py3-none-any.whl size=334491 sha256=fbcc6fea499c27e7d6d434b3c26d30d07ee1b45552ef2c3416fe5ab30ef6d622\n",
            "  Stored in directory: /root/.cache/pip/wheels/a9/7b/9d/e599e33efb84272c1bd485e3fa8aff68ac727f203c9d3e0d13\n",
            "  Building wheel for connection-pool (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for connection-pool: filename=connection_pool-0.0.3-py3-none-any.whl size=4084 sha256=85e821efff90a17ab6bf0c8825f07b1f41fba431921e4144be9f6de8cb6f4acd\n",
            "  Stored in directory: /root/.cache/pip/wheels/a0/66/fa/93d22c4aed2fcf82f4d7a0ab8587cb3e1918153471360f8e07\n",
            "  Building wheel for stopit (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for stopit: filename=stopit-1.1.2-py3-none-any.whl size=11956 sha256=ceab1e707292f312f3c7bcff0a0699ec711354070609d2321ceb493f717c0b2f\n",
            "  Stored in directory: /root/.cache/pip/wheels/e2/d2/79/eaf81edb391e27c87f51b8ef901ecc85a5363dc96b8b8d71e3\n",
            "Successfully built scquint umap-learn pynndescent session-info snakemake connection-pool stopit\n",
            "Installing collected packages: smmap, fonttools, stdlib-list, pyyaml, pynndescent, plac, matplotlib, gitdb, dpath, yte, umap-learn, toposort, stopit, session-info, retry, ratelimiter, pyro-api, pulp, jinja2, gitpython, datrie, connection-pool, configargparse, anndata, snakemake, scanpy, pyro-ppl, scquint\n",
            "  Attempting uninstall: pyyaml\n",
            "    Found existing installation: PyYAML 3.13\n",
            "    Uninstalling PyYAML-3.13:\n",
            "      Successfully uninstalled PyYAML-3.13\n",
            "  Attempting uninstall: plac\n",
            "    Found existing installation: plac 1.1.3\n",
            "    Uninstalling plac-1.1.3:\n",
            "      Successfully uninstalled plac-1.1.3\n",
            "  Attempting uninstall: matplotlib\n",
            "    Found existing installation: matplotlib 3.2.2\n",
            "    Uninstalling matplotlib-3.2.2:\n",
            "      Successfully uninstalled matplotlib-3.2.2\n",
            "  Attempting uninstall: jinja2\n",
            "    Found existing installation: Jinja2 2.11.3\n",
            "    Uninstalling Jinja2-2.11.3:\n",
            "      Successfully uninstalled Jinja2-2.11.3\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "thinc 7.4.0 requires plac<1.2.0,>=0.9.6, but you have plac 1.3.5 which is incompatible.\n",
            "spacy 2.2.4 requires plac<1.2.0,>=0.9.6, but you have plac 1.3.5 which is incompatible.\n",
            "flask 1.1.4 requires Jinja2<3.0,>=2.10.1, but you have jinja2 3.1.2 which is incompatible.\n",
            "datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.\n",
            "albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you have imgaug 0.2.9 which is incompatible.\u001b[0m\n",
            "Successfully installed anndata-0.8.0 configargparse-1.5.3 connection-pool-0.0.3 datrie-0.8.2 dpath-2.0.6 fonttools-4.33.3 gitdb-4.0.9 gitpython-3.1.27 jinja2-3.1.2 matplotlib-3.5.2 plac-1.3.5 pulp-2.6.0 pynndescent-0.5.7 pyro-api-0.1.2 pyro-ppl-1.8.1 pyyaml-6.0 ratelimiter-1.2.0.post0 retry-0.9.2 scanpy-1.9.1 scquint-0.2 session-info-1.0.0 smmap-5.0.0 snakemake-7.8.1 stdlib-list-0.8.0 stopit-1.1.2 toposort-1.7 umap-learn-0.5.3 yte-1.5.1\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "matplotlib",
                  "mpl_toolkits"
                ]
              }
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download processed intron count dat\n",
        "!wget https://figshare.com/ndownloader/files/27696714 -O adata_spl.h5ad"
      ],
      "metadata": {
        "id": "ypi1bs2OTEeV",
        "outputId": "75266b3e-8d95-4f7a-fbf3-e7b862d6a646",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-06-08 03:13:30--  https://figshare.com/ndownloader/files/27696714\n",
            "Resolving figshare.com (figshare.com)... 54.72.163.193, 52.50.42.102, 2a05:d018:1f4:d000:7421:2135:71b2:6a10, ...\n",
            "Connecting to figshare.com (figshare.com)|54.72.163.193|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/27696714/adata_spl.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220608/eu-west-1/s3/aws4_request&X-Amz-Date=20220608T031330Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=fa6a714d2ec76df88d02f677647b3ec25922882f636430fd2cc20aa9fb69d172 [following]\n",
            "--2022-06-08 03:13:30--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/27696714/adata_spl.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220608/eu-west-1/s3/aws4_request&X-Amz-Date=20220608T031330Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=fa6a714d2ec76df88d02f677647b3ec25922882f636430fd2cc20aa9fb69d172\n",
            "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.46.96\n",
            "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.46.96|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 286049939 (273M) [application/octet-stream]\n",
            "Saving to: ‘adata_spl.h5ad’\n",
            "\n",
            "adata_spl.h5ad      100%[===================>] 272.80M  30.5MB/s    in 9.7s    \n",
            "\n",
            "2022-06-08 03:13:41 (28.3 MB/s) - ‘adata_spl.h5ad’ saved [286049939/286049939]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "pQMBiHyKSnCw",
        "outputId": "ad44cf61-464b-417a-9ebf-e4d4d7c9a22c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
            "  import pandas.util.testing as tm\n"
          ]
        }
      ],
      "source": [
        "import anndata\n",
        "import numpy as np\n",
        "\n",
        "from scquint.differential_splicing import run_differential_splicing"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "SW-KOMk0SnCz",
        "outputId": "b545eee4-8556-486c-8ba9-df2866662bfe",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 44518 × 29965\n",
              "    obs: 'FACS.selection', 'age', 'cell', 'cell_ontology_class', 'cell_ontology_id', 'free_annotation', 'method', 'mouse.id', 'sex', 'subtissue', 'tissue', 'n_genes', 'n_counts', 'louvain', 'leiden', 'cell_type', 'plate_id'\n",
              "    var: 'chromosome', 'start', 'end', 'cluster', 'left_exon_count', 'left_genes', 'right_exon_count', 'right_genes', 'genes', 'n_genes', 'genes_cluster', 'n_genes_cluster', 'annotated', 'gene_id', 'original_cluster', 'n_start', 'strand', 'cluster_type'"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ],
      "source": [
        "adata = anndata.read_h5ad(\"adata_spl.h5ad\")\n",
        "adata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "MzP51gkpSnC0",
        "outputId": "994f7386-4e0f-46d2-a26d-2f8df80fd041",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "fibroblast of cardiac tissue           2118\n",
              "endothelial cell of coronary artery    1179\n",
              "monocyte                                490\n",
              "endocardial cell                        164\n",
              "atrial myocyte                          159\n",
              "valve cell                              158\n",
              "smooth muscle cell                       65\n",
              "ventricular myocyte                      46\n",
              "T cell                                   37\n",
              "B cell                                   15\n",
              "Name: cell_ontology_class, dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ],
      "source": [
        "adata.obs[adata.obs.tissue==\"Heart\"].cell_ontology_class.value_counts()[:10]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "WjL-a2edSnC1",
        "outputId": "dd0e97b0-e721-40eb-cc19-edd097fb4dcf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(164, 159)"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "# Specify the indices of the two groups to be compared\n",
        "cell_idx_a = np.where((adata.obs.tissue==\"Heart\") & (adata.obs.cell_ontology_class==\"endocardial cell\"))[0]\n",
        "cell_idx_b = np.where((adata.obs.tissue==\"Heart\") & (adata.obs.cell_ontology_class==\"atrial myocyte\"))[0]\n",
        "len(cell_idx_a), len(cell_idx_b)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "qDmgUEBFSnC2",
        "outputId": "2011f44c-2045-428b-ba18-83414b711800",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "sample sizes:  164 159\n",
            "(323, 29965)\n",
            "filter_min_cells_per_feature\n",
            "filter_singletons\n",
            "(323, 271)\n",
            "filter_min_global_proportion\n",
            "filter_singletons\n",
            "(323, 271)\n",
            "filter_min_cells_per_cluster\n",
            "filter_singletons\n",
            "filter_min_cells_per_cluster\n",
            "filter_singletons\n",
            "(323, 169)\n",
            "Number of intron clusters:  84\n",
            "Number of introns:  169\n",
            "Testing intron cluster  0\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
            "  self[name] = value\n",
            "/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
            "  self[name] = value\n",
            "/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
            "  self[name] = value\n",
            "/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
            "  self[name] = value\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Done\n"
          ]
        }
      ],
      "source": [
        "clusters, introns = run_differential_splicing(\n",
        "    adata, cell_idx_a, cell_idx_b, min_cells_per_cluster=50, min_total_cells_per_intron=50, n_jobs=1,\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "33CR5lWHSnC2",
        "outputId": "2bf9fbc7-1e55-4fa2-babf-3948b493c101",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 237
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "              p_value              ll_null                   ll  n_classes  \\\n",
              "cluster                                                                      \n",
              "0        2.783505e-47  -358.71158349144565  -254.41451957203446          2   \n",
              "40       5.241887e-41   -304.4620056272593  -214.54001975380993          2   \n",
              "20       4.590732e-39  -452.07306917411984  -366.59853066480963          2   \n",
              "33       1.121661e-36  -250.61923551606782  -170.61056985779743          2   \n",
              "48       1.605253e-30  -118.71941662337022   -52.78925329732763          2   \n",
              "\n",
              "                    gene_id gene_name  max_abs_delta_psi  max_abs_lfc_psi  \\\n",
              "cluster                                                                     \n",
              "0        ENSMUSG00000061904        NA           0.803164        29.571781   \n",
              "40       ENSMUSG00000032366        NA           0.803828         3.785967   \n",
              "20       ENSMUSG00000025781        NA           0.723182         2.771993   \n",
              "33       ENSMUSG00000008892        NA           0.798089         6.172689   \n",
              "48       ENSMUSG00000018845        NA           0.900812         5.421720   \n",
              "\n",
              "         ranking   p_value_adj  \n",
              "cluster                         \n",
              "0              0  2.338144e-45  \n",
              "40             1  2.201593e-39  \n",
              "20             2  1.285405e-37  \n",
              "33             3  2.355488e-35  \n",
              "48             4  2.696825e-29  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-aa42b9f8-a59a-4f2f-ad5c-2253112ba3bb\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>p_value</th>\n",
              "      <th>ll_null</th>\n",
              "      <th>ll</th>\n",
              "      <th>n_classes</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_name</th>\n",
              "      <th>max_abs_delta_psi</th>\n",
              "      <th>max_abs_lfc_psi</th>\n",
              "      <th>ranking</th>\n",
              "      <th>p_value_adj</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cluster</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>2.783505e-47</td>\n",
              "      <td>-358.71158349144565</td>\n",
              "      <td>-254.41451957203446</td>\n",
              "      <td>2</td>\n",
              "      <td>ENSMUSG00000061904</td>\n",
              "      <td>NA</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>29.571781</td>\n",
              "      <td>0</td>\n",
              "      <td>2.338144e-45</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>5.241887e-41</td>\n",
              "      <td>-304.4620056272593</td>\n",
              "      <td>-214.54001975380993</td>\n",
              "      <td>2</td>\n",
              "      <td>ENSMUSG00000032366</td>\n",
              "      <td>NA</td>\n",
              "      <td>0.803828</td>\n",
              "      <td>3.785967</td>\n",
              "      <td>1</td>\n",
              "      <td>2.201593e-39</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>4.590732e-39</td>\n",
              "      <td>-452.07306917411984</td>\n",
              "      <td>-366.59853066480963</td>\n",
              "      <td>2</td>\n",
              "      <td>ENSMUSG00000025781</td>\n",
              "      <td>NA</td>\n",
              "      <td>0.723182</td>\n",
              "      <td>2.771993</td>\n",
              "      <td>2</td>\n",
              "      <td>1.285405e-37</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>1.121661e-36</td>\n",
              "      <td>-250.61923551606782</td>\n",
              "      <td>-170.61056985779743</td>\n",
              "      <td>2</td>\n",
              "      <td>ENSMUSG00000008892</td>\n",
              "      <td>NA</td>\n",
              "      <td>0.798089</td>\n",
              "      <td>6.172689</td>\n",
              "      <td>3</td>\n",
              "      <td>2.355488e-35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>1.605253e-30</td>\n",
              "      <td>-118.71941662337022</td>\n",
              "      <td>-52.78925329732763</td>\n",
              "      <td>2</td>\n",
              "      <td>ENSMUSG00000018845</td>\n",
              "      <td>NA</td>\n",
              "      <td>0.900812</td>\n",
              "      <td>5.421720</td>\n",
              "      <td>4</td>\n",
              "      <td>2.696825e-29</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-aa42b9f8-a59a-4f2f-ad5c-2253112ba3bb')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-aa42b9f8-a59a-4f2f-ad5c-2253112ba3bb button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-aa42b9f8-a59a-4f2f-ad5c-2253112ba3bb');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "# clusters has information about all the intron clusters that were tested\n",
        "clusters.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "f7qANOW5SnC3",
        "outputId": "5841b462-8b41-4979-a65e-61f076bc0db6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 299
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  chromosome      start        end  cluster  left_exon_count  \\\n",
              "0      chr10   91119707   91121922        0                9   \n",
              "1      chr10   91119707   91122207        0                9   \n",
              "2      chr10  128491034  128491720        1               11   \n",
              "3      chr10  128491034  128492059        1               11   \n",
              "4      chr11    4702221    4703904        2                4   \n",
              "\n",
              "               left_genes  right_exon_count             right_genes  \\\n",
              "0  ['ENSMUSG00000061904']                 5  ['ENSMUSG00000061904']   \n",
              "1  ['ENSMUSG00000061904']                 2  ['ENSMUSG00000061904']   \n",
              "2  ['ENSMUSG00000090841']                 1  ['ENSMUSG00000090841']   \n",
              "3  ['ENSMUSG00000090841']                14  ['ENSMUSG00000090841']   \n",
              "4  ['ENSMUSG00000059534']                 2  ['ENSMUSG00000059534']   \n",
              "\n",
              "                    genes  n_genes  ... n_start  strand     cluster_type  \\\n",
              "0  ['ENSMUSG00000061904']        1  ...       1       -  shared_acceptor   \n",
              "1  ['ENSMUSG00000061904']        1  ...       1       -  shared_acceptor   \n",
              "2  ['ENSMUSG00000090841']        1  ...       1       -  shared_acceptor   \n",
              "3  ['ENSMUSG00000090841']        1  ...       1       -  shared_acceptor   \n",
              "4  ['ENSMUSG00000059534']        1  ...       1       -  shared_acceptor   \n",
              "\n",
              "          psi_a     psi_b  delta_psi    lfc_psi abs_delta_psi  abs_lfc_psi  \\\n",
              "0  1.000000e+00  0.196836   0.803164   2.344936      0.803164     2.344936   \n",
              "1  6.494021e-12  0.803164  -0.803164 -29.571781      0.803164    29.571781   \n",
              "2  7.666660e-01  0.206015   0.560651   1.895847      0.560651     1.895847   \n",
              "3  2.333340e-01  0.793985  -0.560651  -1.766715      0.560651     1.766715   \n",
              "4  3.145444e-02  0.102419  -0.070964  -1.703142      0.070964     1.703142   \n",
              "\n",
              "   gene_name  \n",
              "0         NA  \n",
              "1         NA  \n",
              "2         NA  \n",
              "3         NA  \n",
              "4         NA  \n",
              "\n",
              "[5 rows x 25 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e7bbbbde-6c5a-4c0b-a908-e058bc15ce4a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chromosome</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>cluster</th>\n",
              "      <th>left_exon_count</th>\n",
              "      <th>left_genes</th>\n",
              "      <th>right_exon_count</th>\n",
              "      <th>right_genes</th>\n",
              "      <th>genes</th>\n",
              "      <th>n_genes</th>\n",
              "      <th>...</th>\n",
              "      <th>n_start</th>\n",
              "      <th>strand</th>\n",
              "      <th>cluster_type</th>\n",
              "      <th>psi_a</th>\n",
              "      <th>psi_b</th>\n",
              "      <th>delta_psi</th>\n",
              "      <th>lfc_psi</th>\n",
              "      <th>abs_delta_psi</th>\n",
              "      <th>abs_lfc_psi</th>\n",
              "      <th>gene_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr10</td>\n",
              "      <td>91119707</td>\n",
              "      <td>91121922</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>5</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>1.000000e+00</td>\n",
              "      <td>0.196836</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>2.344936</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>2.344936</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr10</td>\n",
              "      <td>91119707</td>\n",
              "      <td>91122207</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>2</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>6.494021e-12</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>-0.803164</td>\n",
              "      <td>-29.571781</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>29.571781</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr10</td>\n",
              "      <td>128491034</td>\n",
              "      <td>128491720</td>\n",
              "      <td>1</td>\n",
              "      <td>11</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>1</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>7.666660e-01</td>\n",
              "      <td>0.206015</td>\n",
              "      <td>0.560651</td>\n",
              "      <td>1.895847</td>\n",
              "      <td>0.560651</td>\n",
              "      <td>1.895847</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr10</td>\n",
              "      <td>128491034</td>\n",
              "      <td>128492059</td>\n",
              "      <td>1</td>\n",
              "      <td>11</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>14</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>['ENSMUSG00000090841']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>2.333340e-01</td>\n",
              "      <td>0.793985</td>\n",
              "      <td>-0.560651</td>\n",
              "      <td>-1.766715</td>\n",
              "      <td>0.560651</td>\n",
              "      <td>1.766715</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr11</td>\n",
              "      <td>4702221</td>\n",
              "      <td>4703904</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>['ENSMUSG00000059534']</td>\n",
              "      <td>2</td>\n",
              "      <td>['ENSMUSG00000059534']</td>\n",
              "      <td>['ENSMUSG00000059534']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>3.145444e-02</td>\n",
              "      <td>0.102419</td>\n",
              "      <td>-0.070964</td>\n",
              "      <td>-1.703142</td>\n",
              "      <td>0.070964</td>\n",
              "      <td>1.703142</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 25 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e7bbbbde-6c5a-4c0b-a908-e058bc15ce4a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-e7bbbbde-6c5a-4c0b-a908-e058bc15ce4a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-e7bbbbde-6c5a-4c0b-a908-e058bc15ce4a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ],
      "source": [
        "# introns has information about all the introns corresponding to the tested clusters\n",
        "introns.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "NCfBy1alSnC4",
        "outputId": "5c143387-6f07-4c3a-fa31-ac27681e7b7e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "45"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ],
      "source": [
        "# Number of significant differential splicing events\n",
        "(clusters.p_value_adj < 0.05).sum()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "-gDY_jpbSnC5",
        "outputId": "9cc40ffc-34f7-4ea7-913e-e256ceb2ade8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 205
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  chromosome     start       end  cluster  left_exon_count  \\\n",
              "0      chr10  91119707  91121922        0                9   \n",
              "1      chr10  91119707  91122207        0                9   \n",
              "\n",
              "               left_genes  right_exon_count             right_genes  \\\n",
              "0  ['ENSMUSG00000061904']                 5  ['ENSMUSG00000061904']   \n",
              "1  ['ENSMUSG00000061904']                 2  ['ENSMUSG00000061904']   \n",
              "\n",
              "                    genes  n_genes  ... n_start  strand     cluster_type  \\\n",
              "0  ['ENSMUSG00000061904']        1  ...       1       -  shared_acceptor   \n",
              "1  ['ENSMUSG00000061904']        1  ...       1       -  shared_acceptor   \n",
              "\n",
              "          psi_a     psi_b  delta_psi    lfc_psi abs_delta_psi  abs_lfc_psi  \\\n",
              "0  1.000000e+00  0.196836   0.803164   2.344936      0.803164     2.344936   \n",
              "1  6.494021e-12  0.803164  -0.803164 -29.571781      0.803164    29.571781   \n",
              "\n",
              "   gene_name  \n",
              "0         NA  \n",
              "1         NA  \n",
              "\n",
              "[2 rows x 25 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3fac53b2-15fb-4497-a504-af529f6b98be\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chromosome</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>cluster</th>\n",
              "      <th>left_exon_count</th>\n",
              "      <th>left_genes</th>\n",
              "      <th>right_exon_count</th>\n",
              "      <th>right_genes</th>\n",
              "      <th>genes</th>\n",
              "      <th>n_genes</th>\n",
              "      <th>...</th>\n",
              "      <th>n_start</th>\n",
              "      <th>strand</th>\n",
              "      <th>cluster_type</th>\n",
              "      <th>psi_a</th>\n",
              "      <th>psi_b</th>\n",
              "      <th>delta_psi</th>\n",
              "      <th>lfc_psi</th>\n",
              "      <th>abs_delta_psi</th>\n",
              "      <th>abs_lfc_psi</th>\n",
              "      <th>gene_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr10</td>\n",
              "      <td>91119707</td>\n",
              "      <td>91121922</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>5</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>1.000000e+00</td>\n",
              "      <td>0.196836</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>2.344936</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>2.344936</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr10</td>\n",
              "      <td>91119707</td>\n",
              "      <td>91122207</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>2</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>['ENSMUSG00000061904']</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>-</td>\n",
              "      <td>shared_acceptor</td>\n",
              "      <td>6.494021e-12</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>-0.803164</td>\n",
              "      <td>-29.571781</td>\n",
              "      <td>0.803164</td>\n",
              "      <td>29.571781</td>\n",
              "      <td>NA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2 rows × 25 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3fac53b2-15fb-4497-a504-af529f6b98be')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3fac53b2-15fb-4497-a504-af529f6b98be button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3fac53b2-15fb-4497-a504-af529f6b98be');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ],
      "source": [
        "# The introns belonging to the intron cluster with the lowest p-value\n",
        "introns[introns.cluster==0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kxQuOpIxSnC6"
      },
      "outputs": [],
      "source": [
        "# The results can be visualized with \n",
        "# the cellxgene browser https://scquint.ds.czbiohub.org/tabula-muris/\n",
        "# and UCSC Genome Browser https://genome.ucsc.edu/s/gbenegas/tabulamuris"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.6.10"
    },
    "colab": {
      "name": "differential_splicing_example.ipynb",
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}